File: src/main/java/com/hazza/LeetCodeByJava/SegmentTwo/No53MaximumSubarray.java
package com.hazza.LeetCodeByJava.SegmentTwo;

/**
 * Created with IntelliJ IDEA.
 * Description: https://leetcode.com/problems/maximum-subarray/description/
 * Author: HazzaCheng
 * Contact: hazzacheng@gmail.com
 * Date: 2017-08-09
 * Time: 8:45 AM
 */
public class No53MaximumSubarray {
    //solution 1
    public int maxSubArray1(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int len = nums.length, max = nums[0];
        int dp[] = new int[len];
        dp[0] = max;
        for (int i = 1; i < len; i++) {
            dp[i] = nums[i] + (dp[i - 1] <= 0 ? 0 : dp[i - 1]);
            max = Math.max(max, dp[i]);
        }

        return max;
    }

    //solution 2
    public int maxSubArray(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int len = nums.length, curMax = nums[0], max = nums[0];
        for (int i = 1; i < len; i++) {
            curMax = Math.max(curMax + nums[i], nums[i]);
            max = Math.max(curMax, max);
        }

        return max;
    }

}

File: src/main/python/com/hazza/LeetCodeByPython/SegmentTwo/No53_Maximum_Subarray.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Created by HazzaCheng on 2019/5/21 21:19
# https://leetcode.com/problems/maximum-subarray/
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        curMax, totalMax = nums[0], nums[0]

        for n in nums[1:]:
            curMax = max(curMax + n, n)
            totalMax = max(curMax, totalMax)

        return totalMax

if __name__ == '__main__':
    solution = Solution()

    assert solution.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]) == 6